kiosk/src/Components/AddingGame.tsx

```diff
@@ -58,60 +61,82 @@ const AddingGame: React.FC<IProps> = ({ kiosk }) => {
+                reject();
+                setRenderQRCode(false);
+                pollingStarted.current = false;
+            }, timeoutDuration)
+
+            await getGameCode();
+        });
+    }
+
     useEffect(() => {
-        async function generateKioskCode() {
+        const generateKioskCode = async () => {
             try {
                 const newKioskCode: string = await generateKioskCodeAsync();
                 setKioskCode(newKioskCode);
                 try {
-                    await addGameToKiosk(newKioskCode);
+                    await pollForGames(newKioskCode);
                 } catch (error) {
-                    console.log("Unable to add game to kioksk");
+                    console.log("the promise rejected");
+                    // TODO: dispatch error for time out here
                 }
             }
             catch (error) {
-                setRenderQRCode(false);
                 throw new Error("Unable to generate kiosk code");
+                //TODO: too many set here
             }
         }
 
-        generateKioskCode();
+        if (pollingStarted.current && renderQRCode) {
+            generateKioskCode();
+        }
 
-        function addGameToKiosk(kioskCode: string) {
-            const timeoutDuration = 600000; // wait for 10 minutes until the kiosk code expires
-            const whenToPoll = 5000; // wait for 5 seconds to poll for game code data
-            if (kiosk.state !== KioskState.AddingGame) {
-                return;
+        return () => {
+            pollingStarted.current = true;
+            if (pollFrequency) {
+                clearTimeout(pollFrequency);
             }
-            return new Promise<void>(async (resolve, reject) => {
-                let pollFrequency: any;
-                let pollTimeout: any;
-                const getGameCode = async () => {
-                    try {
-                        const gameCode: string = await getGameCodeAsync(kioskCode);
-                        await kiosk.saveNewGameAsync(gameCode);
-                        tickEvent("kiosk.gameUploaded");
-                        clearTimeout(pollFrequency);
-                        clearTimeout(pollTimeout);
-                        resolve();
-                        kiosk.launchGame(gameCode);
-                    }
-                    catch (error) {
-                        pollFrequency = setTimeout(async () => {
-                            await getGameCode();
-                        }, whenToPoll)
-                    }
-                };
 
-                pollTimeout = setTimeout(() => {
-                    clearTimeout(pollTimeout);
-                    clearTimeout(pollFrequency);
-                    reject();
-                }, timeoutDuration)
-
-                await getGameCode();
-            });
+            if (pollTimeout) {
+                clearTimeout(pollTimeout);
+            }
         }
+
     }, [renderQRCode]);
```

Comment:
This effect might have dependency problem. `pollForGames` will be recreated on each render, and so will `pollFrequency` and `pollTimeout`. They'll be captured here in the effect, but `pollForGames` doesn't have the same life cycle as this effect, and may start acting on different instances of these fields. If you can, it would simplify things to put the timers and `pollForGames` itself inside this effect.

Reply:
I recommend enabling two eslint rules to help catch these kinds of issues:
* react-hooks/rules-of-hooks
* react-hooks/exhaustive-deps

But scrutinize their output. It can be very helpful, but also wrong in some cases.

Reply:
https://reactjs.org/docs/hooks-rules.html

Reply:
Oh, this is good to know. I think this has actually been a large reason as to why the regeneration isn't as smooth as it could be, even with the changes that I've made. 

Does this mean that the timeouts should be in the top-level scope of the use effect? Before, the polls were technically inside the `useEffect`, but very nested (in the `addGameToKiosk` [function](https://github.com/microsoft/pxt-arcade/blob/master/kiosk/src/Components/AddingGame.tsx#L86) that has been renamed to `pollForGames` to better reflect what's happening). I'm guessing this would still mean that the polls would still have different life cycles, thus causing problems.

Reply:
If you moved `pollFrequency`, `pollTimeout`, and `pollForGames` to the top level scope of the use effect, it should clarify the lifetime of these elements, and sync them to the lifecycle of the hook. I can't say for certain whether this is causing an issue though. Looking at it more closely, even though pollForGames and the timer handles are likely getting recreated more frequently than the hook, it seems that only the timers captured by the hook are ever acted on.